bpfd/src/bpf.rs

```diff
@@ -894,4 +942,237 @@ impl BpfManager {
+            // make sure no other eBPF programs are referencing the maps before
+            // allowing it to be deleted.
+            if let Some(map) = self.maps.get_mut(&id.to_string()) {
+                if map.used_by.len() > 1 {
+                    // USed by more than one eBPF program (one would be this program),
+                    // so block the unload.
+                    return false;
+                }
+            }
+        }
+
+        true
+    }
+
+    // This function cleans up a map entry when an eBPF program is
+    // being unloaded. If the eBPF program is the map owner, then
+    // the map is removed from the hash table and the associated
+    // directory is removed. If this eBPF program is referencing a
+    // map from another eBPF program, then this eBPF programs UUID
+    // is removed from the UsedBy array.
+    async fn delete_map(
+        &mut self,
+        id: Uuid,
+        map_prog_uuid: Option<String>,
+    ) -> Result<(), BpfdError> {
+        let (_, map_index) = get_map_index(id, map_prog_uuid.clone());
+
+        if let Some(map) = self.maps.get_mut(&map_index.clone()) {
+            //for uuid in map.used_by.iter() {
+            //    info!("  used_by: {:?}", uuid);
+            //}
+            if let Some(index) = map
+                .used_by
+                .iter()
+                .position(|value| *value == id.to_string())
+            {
+                map.used_by.swap_remove(index);
+            }
+
+            if map.used_by.is_empty() {
+                let (_, path) = calc_map_pin_path(id, map_prog_uuid.clone());
+                self.maps.remove(&map_index.clone());
+                fs::remove_dir_all(path)
+                    .await
+                    .map_err(|e| BpfdError::Error(format!("can't delete map dir: {e}")))?;
+            }
+        } else {
+            return Err(BpfdError::Error("map_pin_path does not exists".to_string()));
+        }
+
+        Ok(())
+    }
+
+    fn rebuild_map_entry(&mut self, id: Uuid, map_prog_uuid: Option<String>) {
+        let (_, map_index) = get_map_index(id, map_prog_uuid.clone());
+
+        if let Some(map) = self.maps.get_mut(&map_index) {
+            map.used_by.push(id.to_string());
+        } else {
+            let (_, map_pin_path) = calc_map_pin_path(id, map_prog_uuid.clone());
+            let map = BpfMap {
+                map_pin_path: map_pin_path.clone(),
+                used_by: vec![id.to_string()],
+            };
+            self.maps.insert(id.to_string(), map);
+        }
+    }
+}
+
+// map_index is a UUID. It is either the programs UUID, or the UUID
+// of another program that map_prog_uuid references.
+// This function also returns a bool, which indicates if the input UUID
+// is the owner of the map (map_prog_uuid is not set) or not (map_prog_uuid
+// is set so the eBPF program is referencing another eBPF programs maps).
+fn get_map_index(id: Uuid, map_prog_uuid: Option<String>) -> (bool, String) {
```

Comment:
I feel like enforcing map_prog_uuid to be an Option<Uuid> would help mimize the to_string calls as well as an extra layer of validation 

if `Uuid::parse_str(<Map_prog_uuid>).unwrap();` fails we know the user entered something wonky

Reply:
Not sure what you mean here?

I did changed it to be a Option<Uuid> as part of the global change to make all Uuids use the actual Uuid type instead of String.

Reply:
That's what I meant, With that we get built in validation 

say a user enters `"ef71d42c-aa21-ERROR-a697-82391d801a81"` by accident in `map_prog_uuid` it will barf 👍

Reply:
I see the problem now, the '<'UUID'>' didn't show up in your comment (or mine) so I thought you were asking for an Option, which it already was. I did make this an Option'<'UUID'>' , so done.